books/kestrel/apt/utilities/find-a-base-case.lisp

```diff
@@ -0,0 +1,154 @@
+;; Author: Grant Jurgensen (grant@kestrel.edu)
+
+(in-package "ACL2")
+
+(include-book "std/util/bstar" :dir :system)
+(include-book "kestrel/utilities/forms" :dir :system)
+(include-book "kestrel/terms-light/expr-calls-fn" :dir :system)
+(include-book "kestrel/utilities/translate" :dir :system)
+(include-book "kestrel/utilities/magic-macroexpand" :dir :system)
+(local (include-book "kestrel/typed-lists-light/pseudo-term-listp" :dir :system))
+(local (include-book "kestrel/typed-lists-light/symbol-listp" :dir :system))
+
+(defthm pseudo-termp-of-if-condition
+  (implies (and (pseudo-termp term)
+                (eq 'if (ffn-symb term)))
+           (pseudo-termp (farg1 term))))
+
+(defthm pseudo-termp-of-if-then-branch
+  (implies (and (pseudo-termp term)
+                (eq 'if (ffn-symb term)))
+           (pseudo-termp (farg2 term))))
+
+(defthm pseudo-termp-of-if-else-branch
+  (implies (and (pseudo-termp term)
+                (eq 'if (ffn-symb term)))
+           (pseudo-termp (farg3 term))))
+
+;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+
+;; Variant on translated terms
+
+;; `term` is the term to find a base-case within.
+;; `fns` is a list of functions to avoid. A base-case must not include these functions.
+;; If `prefer-then` is `t`, we search with a bias toward "then" branches. If it
+;; is `nil`, we are biased toward "else" branches.
+
+;; Returns `(mv erp all-base largest)`, where `erp` is `nil` when a base-case
+;; is found, `all-base` is `t` when the entire term is a base-case, and
+;; if `erp` and `all-base` are both `nil`, then `largest` is the largest base-case.
+(defun find-a-base-case-translated-aux (term fns prefer-then)
+  (declare (xargs :guard (and (pseudo-termp term)
+                              (symbol-listp fns)
+                              (booleanp prefer-then))))
+  (b* (((unless (consp term)) (mv nil t nil))
+       ((unless (eq 'if (ffn-symb term)))
```

Comment:
It may be cleaner to have an explicit case/check for a quoted constant, between the previous check (for variables) and this check (where the use of `ffn-symb` suggests that we are in the case of a call term.